packages/utils/telemetry-utils/src/batchedTelemetryHelper.ts

```diff
@@ -0,0 +1,92 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import type { IDisposable } from "@fluidframework/core-interfaces";
+
+import type {
+	ITelemetryGenericEventExt,
+	ITelemetryLoggerExt,
+	ITelemetryPerformanceEventExt,
+} from "./telemetryTypes.js";
+
+/**
+ * TODO
+ */
+interface ICustomDataMap {
+	[key: string]: number;
+}
+
+/**
+ * TODO
+ */
+export class BatchedTelemetryHelper implements IDisposable {
+	disposed: boolean = false;
+
+	// Stores value of the custom data passed into the logger.
+	private readonly customDataMap: Map<string, number> = new Map<string, number>();
+	// Counter to keep track of the number of times the log function is called.
+	private counter = 0;
+
+	/**
+	 * TODO
+	 */
+	public constructor(
+		private readonly eventBase: ITelemetryGenericEventExt,
+		private readonly logger: ITelemetryLoggerExt,
+		private readonly threshold: number,
+	) {}
+
+	private incrementThresholdCount(): void {
+		this.counter++;
+	}
+
+	private resetThresholdCount(): void {
+		this.counter = 0;
+	}
+
+	private isAboveThreshold(): boolean {
+		return this.counter >= this.threshold;
+	}
+
+	/**
+	 * TODO
+	 */
+	public log(data: ICustomDataMap): void {
+		for (const key of Object.keys(data)) {
+			if (Object.prototype.hasOwnProperty.call(this.customDataMap, key)) {
+				this.customDataMap[key] += data[key];
+			} else {
+				this.customDataMap[key] = data[key];
+			}
+		}
+
+		this.incrementThresholdCount();
+
+		if (this.isAboveThreshold()) {
+			this.sendData();
+		}
+	}
+
+	public sendData(): void {
+		const customData = Object.fromEntries(
+			[...this.customDataMap.entries()].map(([key, value]) => [key, value / this.counter]),
+		);
+
+		// TODO: Add `average` name to the custom data.
+		const telemetryEvent: ITelemetryPerformanceEventExt = {
+			...this.eventBase,
+			...customData,
+		};
+
+		this.logger.sendPerformanceEvent(telemetryEvent);
+		this.resetThresholdCount();
+	}
+
+	public dispose(error?: Error | undefined): void {
+		// TODO: Implement dispose method.
+		// Not sure if calling `sendData()` for each entries (like `SampledTeelmetryHelper` does) is necessary.
+		// We wish to accumulate the data over time and send it every `threshold` times.
+		// Maybe instead `this.customMap.delete()` for each entries?
+	}
```

Comment:
I'd say file a follow-up task for it, and remove the method (and the `disposed` member and `implement IDisposable`) for now. It is nice to have, but unless the "owner" of the helper instance actually sets up things correctly and calls dispose at some point, it doesn't matter much. I don't remember if we have that set up for SampledTelemetryHelper. Since it's not critical for the helper to work, I'd push it to after the GA release.

Reply:
https://github.com/microsoft/FluidFramework/blob/109d5a951e13bfa12a31680a5653c7536a824de8/packages/utils/telemetry-utils/src/sampledTelemetryHelper.ts#L147

Good point. I will remove the `implements IDisposable` but create a follow-up task to implement it in the future. 

+

The `SampledTelemetryHelper` does implement `IDisposable` therefore has a `dispose()` method.

Reply:
https://dev.azure.com/fluidframework/internal/_workitems/edit/8202